### ISSUE TITLE
[BUG] Unable to get VLM to recognize image due to processor misalignment

### ISSUE THREAD

**Issue by nickvdw** [huggingface/smolagents#841]
**Describe the bug**
I'm unable to get a VLM, specificially HuggingFaceTB/SmolVLM-Instruct, to describe the content of an image due to misalignment between the `processor` usage in Smolagent's `TransformersModel` and the "intended" usage of the `processor`. It seems that the `processor` usage in Smolagent's `TransformersModel` may only work for specific VLMs instead of all VLMs.

**Code to reproduce the error**
First, to demonstrate what I would expect as output for a given prompt, I run inference on [HuggingFaceTB/SmolVLM-Instruct](https://huggingface.co/HuggingFaceTB/SmolVLM-Instruct) based on the "How To Get Started" code.

```python
import torch
import requests

from io import BytesIO
from PIL import Image
from transformers import AutoProcessor, AutoModelForVision2Seq

DEVICE = "cuda" if torch.cuda.is_available() else "cpu"

# Load images
url = "https://upload.wikimedia.org/wikipedia/commons/e/e8/The_Joker_at_Wax_Museum_Plus.jpg"
response = requests.get(url)
image = Image.open(BytesIO(response.content)).convert("RGB")

# Initialize processor and model.
processor = AutoProcessor.from_pretrained("HuggingFaceTB/SmolVLM-Instruct")
model = AutoModelForVision2Seq.from_pretrained(
    "HuggingFaceTB/SmolVLM-Instruct",
    torch_dtype=torch.bfloat16,
).to(DEVICE)

# Create input messages
messages = [
    {
        "role": "user",
        "content": [
            {"type": "image"},
            {"type": "text", "text": "Describe the image."}
        ]
    },
]

# Prepare inputs.
inputs = processor.apply_chat_template(messages, add_generation_prompt=True)
inputs = processor(text=inputs, images=[image], return_tensors="pt")
inputs = inputs.to(DEVICE)

# Generate outputs.
generated_ids = model.generate(**inputs, max_new_tokens=500)
generated_texts = processor.batch_decode(
    generated_ids,
    skip_special_tokens=True,
)

print(generated_texts[0])
```
Executing this code cell produces the following output:
```text
Assistant: The image depicts a person wearing a costume, likely a character from a play or a movie. The person is dressed in a purple suit with a yellow shirt underneath. The costume includes a purple bow tie and a purple jacket. The person's face is painted white, with exaggerated makeup, including dark eyebrows, red lips, and black eyeliner. The person is smiling broadly, with their mouth open and teeth exposed. The background of the image is a plain, light-colored wall, which provides a neutral backdrop for the costume.

The person's costume is reminiscent of the Joker, a well-known villain from the Batman comic books and movies. The Joker is known for his distinctive appearance, which includes white face paint, dark hair, and a purple suit. The costume in the image closely resembles the Joker's traditional attire, with the purple suit and yellow shirt providing a striking contrast.

The person's expression is playful and mischievous, suggesting that they are portraying the Joker character. The exaggerated makeup and costume are designed to evoke a sense of comic book villainy, with the exaggerated smile and white face paint adding to the exaggerated appearance.

In summary, the image depicts a person dressed in a Joker-inspired costume, with a white face, dark eyebrows, red lips, and a purple bow tie. The costume includes a purple suit, yellow shirt, and purple jacket, and the person is smiling broadly. The background is a plain, light-colored wall, which provides a neutral backdrop for the costume. The image is likely taken in a studio or a controlled environment, as the lighting is consistent and there are no other elements that might distract from the costume.

This description can be used to answer any questions related to the image, such as identifying the character portrayed, the details of the costume, and the overall mood or tone of the image.
```
Great, this looks as something that I'd expect. Now, I try something similar with Smolagent's `TransformersModel` wrapped by the `CodeAgent`.

```python
from smolagents import (
    TransformersModel,
    CodeAgent,
)

model = TransformersModel(
    model_id="HuggingFaceTB/SmolVLM-Instruct", #"LanguageBind/Video-LLaVA-7B-hf",   #"Qwen/Qwen2.5-VL-3B-Instruct",  # "Qwen/Qwen2-VL-2B-Instruct",  # "Qwen/Qwen2.5-VL-3B-Instruct",
    torch_dtype=torch.bfloat16,
    device_map="cuda",
    max_new_tokens=500,
)

agent = CodeAgent(
    model=model,
    tools=[],
)

agent.run(
    "Describe the image.",
    images=[image],
    max_steps=3,
)
```

Executing this code cell produces the following output:

![Image](https://github.com/user-attachments/assets/672e80a5-ce73-4216-b1c2-b762dcecd7a2)

Note that subsequent steps produce similar outputs.

I dove a bit deeper into the code and it seems that the model is extrapolating based on the default system prompt, which seems to cover similar examples about this 55 year old lumberjack called John Doe (see, e.g., [https://github.com/huggingface/smolagents/blob/main/src/smolagents/prompts/code_agent.yaml#L22-L27](https://github.com/huggingface/smolagents/blob/main/src/smolagents/prompts/code_agent.yaml#L22-L27)).

After diving deeper, I inspected the [prompt_tensor](https://github.com/huggingface/smolagents/blob/main/src/smolagents/models.py#L769) that gets fed into the model. This tensor did not contain the pixel values and its corresponding mask that you'd typically find while working with images in VLMs. Hence, it seems that the processing of the image didn't happen correctly. After further inspection, this seems to be the case due to the misalignment between the `processor` in the `TransformersModel` and what is expected by the model.

It is expected to be processed as:
```python
prompt_tensor = processor.apply_chat_template(messages, add_generation_prompt=True, tools=[...])
prompt_tensor = processor(text=prompt_tensor, images=[image], return_tensors="pt")
```
while it gets processed as:
```python
images = [Image.open(image) for image in images] if images else None
prompt_tensor = self.processor.apply_chat_template(
    messages,
    tools=[get_tool_json_schema(tool) for tool in tools_to_call_from] if tools_to_call_from else None,
    return_tensors="pt",
    tokenize=True,
    return_dict=True,
    images=images,
    add_generation_prompt=True if tools_to_call_from else False,
)
```
By modifying how the `processor` is used in the `TransformersModel` (e.g., first apply the chat template on the text and then call the processor on the text and images), I'm able to get a similar response as the one I get when using the model in a standalone manner. However, the generated output is now a textual description of the image without the corresponding Python code blob that we'd expect for the `CodeAgent`.

![Image](https://github.com/user-attachments/assets/91ce03ed-3f1b-40ec-bad4-f8e518a1ab1f)

I have two questions related to the behavior that I observe in this case:
1. How to make Smolagent's `TransformersModel` compatible with any VLM that may use the `processor` differently from how it is currently implement in Smolagents.
2. How to further continue my adventure w.r.t. enabling SmolVLM within a `CodeAgent`; specifically, how to get it to generate a Python blob that returns me the description that it is currently already generating. Could this be due the system prompt not being recognize by the chat template of SmolVLM? Please let me know if anything is unclear regarding this question. I'd be happy to provide further detail if helpful.

**Error logs (if any)**
N/A

**Expected behavior**
I would have expected this application to work out-of-the-box and would be curious how to make it work for a variety of VLMs with different processing steps.

**Packages version:**
I am using a locally installed version of the `main` branch as of 28/02/2025 8:04 PM GMT+1.

**Additional context**
